openquake/server/dbapi.py

```diff
@@ -0,0 +1,389 @@
+        elif placeholder == '?x':
+            self.xargs.append(arg)
+            return self.ph
+        elif placeholder == '?s':
+            self.sargs.append(arg)
+            return '{}'
+
+    def join(self, sep, arg):
+        ls = []
+        for name in sorted(arg):
+            self.sargs.append(name)
+            value = arg[name]
+            if value is None:
+                ls.append('{} IS NULL')
+            else:
+                self.xargs.append(value)
+                ls.append('{}=' + self.ph)
+        return sep.join(ls)
+
+    def match(self, m_templ):
+        templ = self.rx.sub(self, m_templ)
+        return templ.format(*self.sargs), tuple(self.xargs)
+
+
+def match(m_templ, *m_args):
+    """
+    :param m_templ: a meta template string
+    :param m_args: all arguments
+    :returns: template, args
+
+    Here is an example of usage:
+
+    >>> match('SELECT * FROM job WHERE id=?x', 1)
+    ('SELECT * FROM job WHERE id=?', (1,))
+    """
+    if not m_args:
+        return m_templ, ()
+    try:
+        return _Replacer(m_args).match(m_templ)
+    except IndexError:
+        raise ValueError('Incorrect number of ?-parameters in %s, expected %s'
+                         % (m_templ, len(m_args)))
+
+
+class Db(object):
+    """
+    A wrapper over a DB API 2 connection. See the tutorial.
+    """
+    def __init__(self, connect, *args, **kw):
+        self.connect = connect
+        self.args = args
+        self.kw = kw
+        self.local = threading.local()
+
+    @property
+    def conn(self):
+        try:
+            return self.local.conn
+        except AttributeError:
+            self.local.conn = self.connect(*self.args, **self.kw)
+            return self.local.conn
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, etype, exc, tb):
+        if etype:
+            self.conn.rollback()
+        else:
+            self.conn.commit()
+
+    def __call__(self, m_templ, *m_args, **kw):
+        cursor = self.conn.cursor()
+        templ, args = match(m_templ, *m_args)
+        if kw.get('debug'):
+            print(templ, args)
+        try:
+            if args:
+                cursor.execute(templ, args)
+            else:
+                cursor.execute(templ)
+        except Exception as exc:
+            raise exc.__class__('%s: %s %s' % (exc, templ, args))
+        if templ.lstrip().lower().startswith(('select', 'pragma')):
+            rows = cursor.fetchall()
+
+            if kw.get('scalar'):  # scalar query
+                if not rows:
+                    raise NotFound
+                elif len(rows) > 1:
+                    raise TooManyRows('%s, expected 1' % len(rows))
+                elif len(rows[0]) > 1:
+                    raise TooManyColumns('%s, expected 1' % len(rows[0]))
+                return rows[0][0]
+            elif kw.get('one'):  # query returning a single row
+                if not rows:
+                    raise NotFound
```

Comment:
This answers some of my previous questions. I suggest to make it clearer in the documentation.